include/nbl/builtin/hlsl/cmath.hlsl

```diff
@@ -0,0 +1,258 @@
+// Copyright (C) 2022 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+#ifndef _NBL_BUILTIN_HLSL_CMATH_INCLUDED_
+#define _NBL_BUILTIN_HLSL_CMATH_INCLUDED_
+
+#include <nbl/builtin/hlsl/limits.hlsl>
+
+// C++ headers
+#ifndef __HLSL_VERSION
+#include <cmath>
+#endif
+
+namespace nbl
+{
+namespace hlsl
+{
+
+
+#ifdef __cplusplus
+#define INTRINSIC_NAMESPACE(x) std::x
+#else
+#define INTRINSIC_NAMESPACE(x) x
+#endif
+
+#define NBL_ALIAS_BINARY_FUNCTION2(name,impl) template<class T> T name(T x, T y) { return INTRINSIC_NAMESPACE(impl)(x, y); }
+#define NBL_ALIAS_BINARY_FUNCTION(fn) NBL_ALIAS_BINARY_FUNCTION2(fn,fn)
+#define NBL_ALIAS_UNARY_FUNCTION2(name,impl)  template<class T> T name(T x) { return INTRINSIC_NAMESPACE(impl)(x); }
+#define NBL_ALIAS_UNARY_FUNCTION(fn)  NBL_ALIAS_UNARY_FUNCTION2(fn,fn)
+
+#define NBL_ALIAS_FUNCTION_WITH_OUTPUT_PARAM(fn, out_type) \
+template<class T>  \
+T fn(T x, NBL_REF_ARG(out_type) y) { \
+    NBL_LANG_SELECT(out_type, T) out_; \
+    T ret = INTRINSIC_NAMESPACE(fn)(x, NBL_ADDRESS_OF(out_)); \
+    y = out_type(out_); \
+    return ret; \
+}
+
+// Trigonometric functions
+NBL_ALIAS_UNARY_FUNCTION(cos)
+NBL_ALIAS_UNARY_FUNCTION(sin)
+NBL_ALIAS_UNARY_FUNCTION(tan)
+NBL_ALIAS_UNARY_FUNCTION(acos)
+NBL_ALIAS_UNARY_FUNCTION(asin)
+NBL_ALIAS_UNARY_FUNCTION(atan)
+NBL_ALIAS_BINARY_FUNCTION(atan2)
+
+// Hyperbolic functions
+NBL_ALIAS_UNARY_FUNCTION(cosh)
+NBL_ALIAS_UNARY_FUNCTION(sinh)
+NBL_ALIAS_UNARY_FUNCTION(tanh)
+
+template<class T>
+T acosh(T x)
+{
+    return INTRINSIC_NAMESPACE(log)(x + INTRINSIC_NAMESPACE(sqrt)(x*x - T(1)));
+}
+
+template<class T>
+T asinh(T x)
+{
+    return INTRINSIC_NAMESPACE(log)(x + INTRINSIC_NAMESPACE(sqrt)(x*x + T(1)));
+}
+
+template<class T>
+T atanh(T x)
+{
+    return T(0.5) * INTRINSIC_NAMESPACE(log)((T(1)+x)/(T(1)-x));
+}
+
+
+// Exponential and logarithmic functions
+NBL_ALIAS_UNARY_FUNCTION(exp)
+NBL_ALIAS_FUNCTION_WITH_OUTPUT_PARAM(frexp, int32_t)
+NBL_ALIAS_BINARY_FUNCTION(ldexp)
+NBL_ALIAS_UNARY_FUNCTION(log)
+NBL_ALIAS_UNARY_FUNCTION(log10)
+NBL_ALIAS_FUNCTION_WITH_OUTPUT_PARAM(modf, T)
+NBL_ALIAS_UNARY_FUNCTION(exp2)
+NBL_ALIAS_UNARY_FUNCTION(log2)
+NBL_ALIAS_UNARY_FUNCTION2(logb,log)
+
+template<class T> 
+T expm1(T x) 
+{ 
+    return INTRINSIC_NAMESPACE(exp)(x) - T(1); 
+}
+
+template<class T> 
+T log1p(T x) 
+{ 
+    return INTRINSIC_NAMESPACE(log)(x + T(1)); 
+}
```

Comment:
hehehe, no.

These are very special functions and @Erfan-Ahmadi will tell you why.

Reply:
I think we encountered it in bezier arclen computations. 
Anyways that's a special function that's supposed to be more accurate for values near 0 than log(1+x)

Reply:
AFAIK both Boost and MUSL are permissively licensed (notices must be preserved) and Boost is more likely to have specialized implementations for different float sizes
https://www.boost.org/doc/libs/1_77_0/boost/math/special_functions/log1p.hpp

Reply:
Hmm given that `log` functions on GPUs are somewhat accelerated, snooping the implementations of cmath (which OpenCL requires) from clvk, sycl, or LLVM is probably best
https://github.com/llvm-mirror/libclc/blob/master/generic/lib/math/log1p.cl